### PR TITLE
Objective Edit Fixes - bug fix and usability improvements

### PR DESCRIPTION
**Pull request by bethanyaconnor**
Closes [PLAT-453](https://codedotorg.atlassian.net/browse/PLAT-453).

The commits in this PR:
- b65e857fe866780750b8747c407a335045ceaa91 fixes a bug where adding an objective would edit the last objective instead 
- To support saving the text if the user clicks out of the box
  - 1f86a78e88a55cd8531636f3c68f849c41073e0a does some refactor work to support this, particularly adding a new component so that the use only has to click outside of the particular line they're editing
  - 20e4957e2950a79b9d1a7901c38a70b77c4440f3 makes use of `react-onclickoutside` to handle detecting clicks
- 8b8449d8a9908c0d7a024fbddf04823a159e5b74 fixes up some tests

## Links

- [spec]()
- [jira]()

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by bethanyaconnor** [comment]
@davidsbailey and @dmcavoy -- I'm having trouble adding both of you as reviewers on the same PR (I can add one of you but not both), so I'm tagging you here as a makeshift review request

File: apps/src/lib/levelbuilder/lesson-editor/ObjectivesEditor.jsx [hunk 1]
```diff
@@ -51,6 +52,99 @@ const styles = {
   }
 };
 
+const ObjectiveLine = onClickOutside(
```
Comment (dmcavoy) [inline apps/src/lib/levelbuilder/lesson-editor/ObjectivesEditor.jsx]: Can you put ObjectiveLine in its own file so we keep the react files small and easy to look at?

**Comment by dmcavoy** [comment]
Hey Bethany - I started looking at this and Declan woke up. More review coming later :)

**Review by dmcavoy — APPROVED** [review]
Looks good to me after split the components into separate files. 

I personally find videos/gifs really helpful in understanding quickly the actions of what is happening in the code when I review. I think this PR is one of those places that it would be nice to see that action as a reviewer. Just a thought for the future! The code is really clean and clear so I was able to make sense of it without it I just feel like it takes me a little longer.

File: apps/src/lib/levelbuilder/lesson-editor/ObjectivesEditor.jsx [hunk 2]
```diff
@@ -51,6 +52,99 @@ const styles = {
...
+
+    constructor(props) {
+      super(props);
+
+      this.state = {
+        description: props.description
+      };
+    }
+
+    handleClickOutside = evt => {
```
Comment (davidsbailey) [inline apps/src/lib/levelbuilder/lesson-editor/ObjectivesEditor.jsx]: nit: unused param? can just be `() => {` ...

**Review by davidsbailey — APPROVED** [review]
I pulled this down locally and it looks good! I looked at the docs for this library, and I think I am missing the "why you would need this" part of the puzzle. Did you turn to this library because of problematic behavior when trying to manage the events directly? I'm not saying you should change this, but I know we also have code that does similar things by adding/removing event handlers on the document body, so I am curious what your experience was.

**Comment by bethanyaconnor** [comment]
I guess I couldn't find anywhere we handle clicks outside of a component that don't use this library. The closest I could think of was the code for dragging but that seemed overly complex for this need. Do you know of anywhere else?

**Comment by davidsbailey** [comment]
Here's the best one I could find in our codebase: https://github.com/code-dot-org/code-dot-org/blob/94271c22575f186103f061ccfed3ab1a75c0a3ec/apps/src/templates/DropdownButton.js#L74-L92

Here's a popular answer on stack overflow which shows roughly the same approach: https://stackoverflow.com/questions/32553158/detect-click-outside-react-component/42234988#42234988

I guess this makes me think that the non-library approach is actually pretty reliable. However, I really like how your approach cleans up the code, and I don't think we want to write our own wrapper to duplicate what the library does, so I'm going to say thanks for listening and suggest keeping your current approach :-)

**Comment by bethanyaconnor** [comment]
Ahh I think I saw that stack overflow but was having trouble figuring out what could actually be done in React 15 🙃

**Comment by bethanyaconnor** [comment]
@dmcavoy Thanks for the feedback! I'll keep that in mind for the future. :) My new computer is _way_ better at taking screen captures than my old one so I'm still working on making this a habit.

For posterity, here is a capture of what I added in this PR:

https://user-images.githubusercontent.com/46464143/103956947-746abe80-50fe-11eb-9c8d-bf5abe225f1d.mov

File: apps/src/lib/levelbuilder/lesson-editor/ObjectivesEditor.jsx (hunk 1 above)
Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/lesson-editor/ObjectivesEditor.jsx]: Done!

Comment (dmcavoy) [reply apps/src/lib/levelbuilder/lesson-editor/ObjectivesEditor.jsx]: Thank you!

**Pull request merged**